src/Type/IntersectionType.php

```diff
@@ -800,8 +800,30 @@ public function setOffsetValueType(?Type $offsetType, Type $valueType, bool $uni
 
 		$result = $this->intersectTypes(static fn (Type $type): Type => $type->setOffsetValueType($offsetType, $valueType, $unionValues));
 
-		if ($offsetType !== null && $this->isList()->yes() && $this->isIterableAtLeastOnce()->yes() && (new ConstantIntegerType(1))->isSuperTypeOf($offsetType)->yes()) {
```

Comment:
this previously existing logic, which was implemented for https://github.com/phpstan/phpstan/issues/12131 - only worked for index `1`. the new logic works for the generic case